File: packages/capacitor/android/src/main/java/app/eodin/capacitor/EodinCapacitorDeeplinkPlugin.kt
package app.eodin.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import app.eodin.deeplink.EodinDeeplink

@CapacitorPlugin(name = "EodinDeeplink")
class EodinCapacitorDeeplinkPlugin : Plugin() {

    @PluginMethod
    fun configure(call: PluginCall) {
        val apiEndpoint = call.getString("apiEndpoint")
        val service = call.getString("service")

        if (apiEndpoint == null || service == null) {
            call.reject("apiEndpoint and service are required")
            return
        }

        EodinDeeplink.configure(
            context = context,
            apiEndpoint = apiEndpoint,
            service = service
        )
        call.resolve()
    }

    @PluginMethod
    fun checkDeferredParams(call: PluginCall) {
        EodinDeeplink.checkDeferredParams { result ->
            result.fold(
                onSuccess = { params ->
                    val data = JSObject().apply {
                        put("path", params.path)
                        put("resourceId", params.resourceId)
                        put("hasParams", params.hasParams)
                        params.metadata?.let { meta ->
                            val metaObj = JSObject()
                            meta.forEach { (key, value) -> metaObj.put(key, value) }
                            put("metadata", metaObj)
                        }
                    }
                    call.resolve(data)
                },
                onFailure = { error ->
                    call.reject(error.message ?: "Unknown error")
                }
            )
        }
    }

    @PluginMethod
    fun isReady(call: PluginCall) {
        val data = JSObject()
        data.put("ready", EodinDeeplink.isReady)
        call.resolve(data)
    }
}

File: packages/capacitor/ios/Sources/EodinCapacitorPlugin/EodinCapacitorDeeplinkPlugin.swift
import Capacitor

@objc(EodinCapacitorDeeplinkPlugin)
public class EodinCapacitorDeeplinkPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "EodinCapacitorDeeplinkPlugin"
    public let jsName = "EodinDeeplink"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkDeferredParams", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isReady", returnType: CAPPluginReturnPromise),
    ]

    @objc func configure(_ call: CAPPluginCall) {
        guard let apiEndpoint = call.getString("apiEndpoint"),
              let service = call.getString("service") else {
            call.reject("apiEndpoint and service are required")
            return
        }

        EodinDeeplink.configure(apiEndpoint: apiEndpoint, service: service)
        call.resolve()
    }

    @objc func checkDeferredParams(_ call: CAPPluginCall) {
        EodinDeeplink.checkDeferredParams { result in
            switch result {
            case .success(let params):
                var data: [String: Any] = [
                    "hasParams": params.hasParams
                ]
                data["path"] = params.path ?? NSNull()
                data["resourceId"] = params.resourceId ?? NSNull()
                data["metadata"] = params.metadata ?? NSNull()
                call.resolve(data)
            case .failure(let error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func isReady(_ call: CAPPluginCall) {
        call.resolve(["ready": EodinDeeplink.isReady])
    }
}
